130 generate batch data from transform

Nic-Ma opened the pull request:
Fixes #130 .

### Description
This PR mainly contains 4 features:
1. **Generate 1 batch** data from crop transform, for example:
    input 3 images, crop out 4 smaller images on every original images, then generate batch=12.
    All the following transforms can still apply to every item in the batch list(_Compose function_).
2. **General crop** transform at the vanilla level.
    It's a general crop transform, can support "center + size" or "start point + end point" modes.
3. Randomly crop out batch data based on a **positive/negative label ratio**.
    According to the label values, randomly select "crop centers" at positive or negative label positions.
4. Enhancement to default **collate_fn** for **list** data type.

Usage example:
1. Define transforms for dict data:
```py
transforms = [
    LoadNiftid(keys=['image', 'extral', 'label']),
    NormalizeIntensityd(keys=['image'], divdior=1.0),
    RandCropByPosNegLabeld(
        keys=['image', 'extral'],
        label_key='label',
        size=[128, 128, 128],
        pos=1, neg=1, num_samples=2
    ),
    RandRotate90d(keys=['image', 'label'])
    ... ...
]
```
2. Define Dataset, DataLoader and use "list_data_collate":
```py
dataset = NiftiDataset(transforms=transforms)
dataloader = DataLoader(dataset=dataset, batch_size=4, collate_fn=list_data_collate)
```

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

Nic-Ma commented:
Hi @ericspod , @atbenmurray , @wyli  and @yanchengnv ,

Could you please help review this PR?
As we still don't have an example of dict level transforms, I put the usage in the PR description directly.
I will develop a training example when all the dict level transforms & training example are ready.
Thanks.

wyli commented on monai/transforms/composables.py:
```diff
@@ -120,6 +121,61 @@ def __call__(self, data):
+    def __call__(self, data):
+        d = dict(data)
+        label = d[self.label_key]
+        centers = generate_pos_neg_label_crop_centers(label, self.size, self.batch_size, self.pos_ratio, self.R)
```
would be great to keep a record of the `centers` after generating the patches, i think we could make `centers` an instance variable, or make new keys in the output data

Nic-Ma replied:
Hi @wyli ,

Thanks for your comments, I already updated code to save centers and use randomize API.

Nic-Ma commented:
Thanks @wyli for the review and comments.
I also added another test case to testCompose and added detailed docstring to Compose.
Do you guys have any more comments?
Thanks.

wyli commented on monai/transforms/compose.py:
```diff
@@ -86,7 +99,13 @@ def __init__(self, transforms=None):
+            # all the following transforms should apply to every item of the list.
+            if isinstance(input_, list):
+                for i, item in enumerate(input_):
+                    input_[i] = transform(item)
```
potentially losing the capability of processing `enumerate(input_)` with multiple workers... personally I don't quite like this idea of treating the image-patch sampling as a composable transform. But I guess I can approve this PR because it doesn't break any existing features

Nic-Ma replied:
Thanks for your review, I already updated code for all comments.

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -206,6 +206,58 @@ def __call__(self, img):
 
 
+@export
+class GeneralCrop:
```
Fixed in latest code.

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -206,6 +206,58 @@ def __call__(self, img):
+        assert (np.subtract(self.roi_end, self.roi_start) >= 0).all(), 'invalid roi range.'
+        if len(self.roi_start) == 1:
+            return img[:, self.roi_start[0]:self.roi_end[0]].copy()
+        elif len(self.roi_start) == 2:
```
Fixed in latest code.

Nic-Ma commented on monai/transforms/composables.py:
```diff
@@ -120,6 +121,65 @@ def __call__(self, data):
+        batch_size (int): number of samples (crop regions) to take in each batch.
+    """
+
+    def __init__(self, keys, label_key, size, pos=1, neg=1, batch_size=1):
```
Fixed in latest code.

Nic-Ma commented on monai/transforms/utils.py:
```diff
@@ -157,3 +157,54 @@ def one_hot(labels, num_classes):
+    valid_end = np.subtract(max_size + np.array(1), size / np.array(2)).astype(np.uint16)  # add 1 for random
+    # int generation to have full range on upper side, but subtract unfloored size/2 to prevent rounded range
+    # from being too high
+    for i in range(2):  # need this because np.random.randint does not work with same start and end
```
Fixed in latest code.

Nic-Ma commented on monai/transforms/utils.py:
```diff
@@ -157,3 +157,54 @@ def one_hot(labels, num_classes):
     return onehot.reshape(tuple(labels.shape) + (num_classes,)).astype(labels.dtype)
+
+
+def generate_pos_neg_label_crop_centers(label, size, batch_size, pos_ratio, rand_state):
```
Fixed in latest code.

Nic-Ma commented on monai/transforms/utils.py:
```diff
@@ -157,3 +157,54 @@ def one_hot(labels, num_classes):
     return onehot.reshape(tuple(labels.shape) + (num_classes,)).astype(labels.dtype)
+
+
+def generate_pos_neg_label_crop_centers(label, size, batch_size, pos_ratio, rand_state):
```
Fixed in latest code.

wyli commented on monai/transforms/composables.py:
```diff
@@ -120,6 +121,66 @@ def __call__(self, data):
+        for key in data.keys():
+            if key in self.keys:
+                img = d[key]
+                image_samples = list()
```
this is unused...

Nic-Ma replied:
Removed in the latest patch.
Thanks.

wyli commented on monai/transforms/composables.py:
```diff
@@ -120,6 +121,66 @@ def __call__(self, data):
+        for i, center in enumerate(self.centers):
+            cropper = SpatialCrop(roi_center=tuple(center), roi_size=self.size)
+            results[i][self.label_key] = cropper(label)
+        return results
```
in this case the user couldn't choose to not cropping data in `self.label_key`?  I suppose `self.keys` already have everything we want to crop

Nic-Ma replied:
Adjusted the logic and unit test in the latest patch.
Thanks.
